SPEC.md

```diff
@@ -0,0 +1,121 @@
+# CNI Specification
+Note that the prime source of information about CNI is to be considered the reference implementation and test suite.
+This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.
+
+## Core Language
+A CNI document consists of an arbitrary amount of statements.
+A statement may either be a section declaration, key-value pair, or key-rawvalue pair.
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
```

Comment:
The trailing non-vertical whitespace is not part of the value (as it may not end with whitespace).
This does not interact with flexspace (for the same reason).
If you want a value that starts/ends with whitespace, you must use raw values.